model/stats/stats.go

```diff
@@ -0,0 +1,276 @@
+func GenerateDailyTaskStats(projectId string, requester string, day time.Time, tasks []string, jobRunTime time.Time) error {
+	grip.Infof("Generating daily task stats for (%v, %v, %v, %v)", projectId, requester, day, tasks)
+	start := getUTCDay(day)
+	end := start.Add(24 * time.Hour)
+	pipeline := dailyTaskStatsPipeline(projectId, requester, start, end, tasks, jobRunTime)
+	err := aggregateIntoCollection(tasksCollection, pipeline, dailyTaskStatsCollection)
+	if err != nil {
+		return errors.Wrap(err, "Failed to aggregate daily task stats")
+	}
+
+	grip.Infof("Generating daily task stats from old tasks for (%v, %v, %v, %v)", projectId, requester, day, tasks)
+	start = getUTCDay(day)
+	end = start.Add(24 * time.Hour)
+	pipeline = dailyTaskStatsForOldTasksPipeline(projectId, requester, start, end, tasks, jobRunTime)
+	err = aggregateIntoCollection(oldTasksCollection, pipeline, dailyTaskStatsCollection)
+	if err != nil {
+		return errors.Wrap(err, "Failed to aggregate daily task stats")
+	}
+	return nil
+}
+
+//////////////////////////////////////////////////////
+// Functions to find which stats need to be updated //
+//////////////////////////////////////////////////////
+
+type StatsToUpdate struct {
+	ProjectId string    `bson:"project"`
+	Requester string    `bson:"requester"`
+	Hour      time.Time `bson:"date"`
+	Day       time.Time `bson:"day"`
+	Tasks     []string  `bson:"task_names"`
+}
+
+func (s StatsToUpdate) canMerge(other StatsToUpdate) bool {
+	return s.ProjectId == other.ProjectId && s.Requester == other.Requester && s.Hour.UTC() == other.Hour.UTC()
+}
+
+// Returns true if this StatsToUpdate should be sorted before the other.
+func (s StatsToUpdate) lt(other StatsToUpdate) bool {
+	if s.ProjectId < other.ProjectId {
+		return true
+	} else if s.ProjectId == other.ProjectId {
+		if s.Hour.UnixNano() < other.Hour.UnixNano() {
+			return true
+		} else if s.Hour.UnixNano() == other.Hour.UnixNano() {
+			if s.Requester < other.Requester {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Merges two StatsToUpdate.
+// This method does not check that the objects can be merged.
+func (s StatsToUpdate) merge(other StatsToUpdate) StatsToUpdate {
+	tasks := s.Tasks
+	for _, t := range other.Tasks {
+		if !containsTask(tasks, t) {
+			tasks = append(tasks, t)
+		}
+	}
+	return StatsToUpdate{s.ProjectId, s.Requester, s.Hour, s.Day, tasks}
+}
+
+// Indicates if a list of strings contains a specific string.
+func containsTask(tasks []string, task string) bool {
+	for _, t := range tasks {
+		if t == task {
+			return true
+		}
+	}
+	return false
+}
+
+// Find the stats that need to be updated as a result of tasks finishing between 'start' and 'end'.
+// The results are ordered by project id, then hour, then requester.
+func FindStatsToUpdate(projectId string, start time.Time, end time.Time) ([]StatsToUpdate, error) {
+	grip.Infof("Finding tasks that need their stats updated for (%v, %v, %v)", projectId, start, end)
```

Comment:
maybe it'd be better to use a grip.Info(message.Fields{}) rather than a formatted string.